File: 733.flood-fill.py
#
# @lc app=leetcode id=733 lang=python3
#
# [733] Flood Fill
#


# @lc code=start
# TODO: dfs
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        R, C = len(image), len(image[0])
        color = image[sr][sc]
        if color == newColor: return image
        def dfs(r: int, c:int ):
            if image[r][c] == color: 
                image[r][c] = newColor
                if r-1 >=0: dfs(r-1,c)
                if r+1 < R: dfs(r+1,c)
                if c-1 >= 0: dfs(r,c-1)
                if c+1 < C: dfs(r,c+1)
        
        dfs(sr,sc)
        
        return image
        
# @lc code=end

File: graph/02_num_of_islands/733.flood-fill.cpp
/*
 * @lc app=leetcode id=733 lang=cpp
 *
 * [733] Flood Fill
 */

#include "bits/stdc++.h"

using namespace std;

// https://zxi.mytechroad.com/blog/searching/leetcode-733-flood-fill/

// @lc code=start

class Solution {
 public:
  vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc,
                                int newColor) {
    if (image[sr][sc] == newColor) return image;

    int R = image.size(), C = image[0].size();

    dfs(image, sr, sc, R, C, image[sr][sc], newColor);
    return image;
  }

  void dfs(vector<vector<int>>& image, int r, int c, int R, int C, int orgColor,
           int newColor) {
    if (r < 0 || r >= R || c < 0 || c >= C) return;
    // if image[r][c] == color, it means current node indirect or direct connect
    // to source node
    if (image[r][c] != orgColor) return;
    image[r][c] = newColor;
    dfs(image, r - 1, c, R, C, orgColor, newColor);
    dfs(image, r + 1, c, R, C, orgColor, newColor);
    dfs(image, r, c + 1, R, C, orgColor, newColor);
    dfs(image, r, c - 1, R, C, orgColor, newColor);
  }
};

// @lc code=end

int main(int argc, char const* argv[]) {
  auto greeting = [](string name) { cout << "Hello " + name << endl; };
  greeting("Ryan");

  int a = 10;
  cout << a << endl;
  auto inc = [&]() { ++a; };
  inc();
  cout << a << endl;

  return 0;
}
